stepflow-rs/crates/stepflow-execution/src/debug_executor.rs

```diff
@@ -0,0 +1,1177 @@
+        let must_execute_indices: Vec<usize> = self
+            .flow
+            .steps()
+            .iter()
+            .enumerate()
+            .filter(|(_, step)| step.must_execute())
+            .map(|(idx, _)| idx)
+            .collect();
+
+        let mut all_newly_queued = Vec::new();
+        for idx in must_execute_indices {
+            let newly_needed = self.state.add_or_update_needed(idx);
+            for dep_idx in newly_needed.iter() {
+                all_newly_queued.push(self.flow.step(dep_idx).id.clone());
+            }
+        }
+
+        log::debug!(
+            "Queued must_execute steps. Newly discovered: [{}]",
+            all_newly_queued.join(", ")
+        );
+
+        all_newly_queued
+    }
+
+    /// Evaluate a step: queue it, run all dependencies, return the result.
+    ///
+    /// This is idempotent - if the step is already completed, returns the cached result.
+    pub async fn eval_step(&mut self, step_id: &str) -> Result<FlowResult> {
+        let step_index = self
+            .state
+            .step_index_map()
+            .step_index(step_id)
+            .ok_or_else(|| ExecutionError::StepNotFound {
+                step: step_id.to_string(),
+            })?;
+
+        // If already completed, return cached result
+        if let Some(result) = self.state.get_step_result(step_index) {
+            return Ok(result.clone());
+        }
+
+        // Queue the step (discovers dependencies)
+        self.state.add_or_update_needed(step_index);
+
+        // Run until target step completes
+        while !self.state.is_completed(step_index) {
+            let ready = self.state.ready_steps();
+            if ready.is_empty() {
+                return Err(error_stack::report!(ExecutionError::Deadlock)
+                    .attach_printable(format!("Cannot make progress towards step '{}'", step_id)));
+            }
+
+            // Execute all ready steps
+            for idx in ready.iter() {
+                self.execute_step_by_index(idx).await?;
+                // Re-evaluate dependencies after each completion
+                self.state.add_or_update_needed(step_index);
+            }
+        }
+
+        Ok(self.state.get_step_result(step_index).unwrap().clone())
+    }
+
+    /// Evaluate the workflow output: queue dependencies, run them, return resolved output.
+    ///
+    /// This also ensures all `must_execute` steps complete successfully.
+    pub async fn eval_output(&mut self) -> Result<FlowResult> {
+        // Queue output dependencies
+        let output_needs = self.flow.output().needed_steps(&self.state);
+        for idx in output_needs.iter() {
+            self.state.add_or_update_needed(idx);
+        }
+
+        // Run until output is resolvable and all needed steps complete
+        loop {
+            // Check if output is ready
+            let output_needs = self.flow.output().needed_steps(&self.state);
```

Comment:
It feels like a lot of this debug executor is duplicative. I think we may be able to simplify. It seems like the main thing is that while normal execution is "drive everything to compute the output", this is more of a "drive everything to complete what the user has requested". This makes me wonder if we should have the debug executor manage a separate set of "requested" computations, and use that similar to how the `FlowExecutor` drives things. Then all of these are different forms of `drain_requested(...)`. It may not even need to be a queue -- we may be able to just have a method `eval_requested(...)` and then have the different `eval` methods call it.